torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+        has been wrapped with FSDP.
+        """
+        with self.set_state_dict_type(StateDictType.LOCAL_STATE_DICT):
+            return FullyShardedDataParallel.state_dict(self, *args, **kwargs)
+
+    @staticmethod
+    def _pre_load_state_dict_hook(
+        module: nn.Module,
+        state_dict: Union[Dict[str, torch.Tensor], "OrderedDict[str, torch.Tensor]"],
+        prefix: str,
+        *args: Any,
+    ) -> None:
+        """
+        ``_pre_state_dict_hook` is called before ``self._load_from_state_dict()``
+        is called. ``self.state_dict_type`` is used to decide what preprocessing
+        will be done.
+        """
+        self = cast(FlatParameter, module)
+        if self.state_dict_type == StateDictType.STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        elif self.state_dict_type == StateDictType.LOCAL_STATE_DICT:
+            _replace_by_prefix(state_dict, prefix, f"{prefix}_fsdp_wrapped_module.")
+            flat_param = self.module.flat_param
+            key = f"{prefix}_fsdp_wrapped_module.flat_param"
+            load_tensor = state_dict[key]
+            assert isinstance(
+                load_tensor, ShardedTensor
+            ), "Tensors in local_state_dict should be ShardedTensor."
+            shards = load_tensor.local_shards()
+            assert len(
+                shards
+            ), "load_local_state_dict assume one shard per ShardedTensor."
+            load_tensor = cast(torch.Tensor, shards[0].tensor)
+            if flat_param.num_padded not in (0, flat_param.numel()):
+                assert load_tensor.numel() < flat_param.numel(), (
+                    f"Local shard size = {flat_param.numel()} and the tensor in "
+                    f"the state_dict is {load_tensor.numel()}."
+                )
+                load_tensor = F.pad(load_tensor, [0, flat_param.num_padded])
+            state_dict[key] = load_tensor
+        elif self.state_dict_type == StateDictType.SHARDED_STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        else:
+            raise ValueError(f"Unknown StateDictType {self.state_dict_type}.")
+
+    def load_state_dict(
+        self,
+        state_dict: "OrderedDict[str, torch.Tensor]",
+        strict: bool = True,
+    ) -> NamedTuple:
+        """
+        The entry point of all three FSDP load_state_dict APIs.
+        ``self.state_dict_type`` decides which code path to execute.
+
+        .. warning:: This needs to be called on all ranks, since synchronization
+            primitives may be used.
+        """
+        torch.cuda.synchronize()
+        if self.state_dict_type == StateDictType.STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        elif self.state_dict_type == StateDictType.LOCAL_STATE_DICT:
+            return super().load_state_dict(state_dict, strict)
```

Comment:
we use `super()` call here but don't use it in `state_dict` implementation where we actually return `FullyShardedDataParallel.state_dict()`, that seems inconsistent and can cause issues/confusion?

Reply:
local_state_dict() and sharded_state_dict() should all call `self.state_dict`. And self.state_dict has to call super().state_dict(). 

local_state_dict() and sharded_state_dict() are just wrapper to set state_dict_type first then call self.state_dict. The full state_dict does not have to do so because the state_dict_type defaults to STATE_DICT.